### PR TITLE
Fully automated bootstrapping for production and adhoc instances

### PR DESCRIPTION
**Pull request by philbogle**
Developers can now fully bootstrap an adhoc or dashboard dashboard+pegasus instance using just add_frontend_server, without requiring any manual post-install steps.  
- If `-e adhoc` is specified, an _unmonitored_ adhoc instance will be created using the `unmonitored-standlone` role, using a local mysql and redis instance and no NewRelic monitoring.  It will be synced to the staging branch by default but developers can easily change that.  
- If `-e production` is specified, a production instance will be created using the standard `frontend` role.

The changes include the following:
- fixes to the chef recipes to correctly create and seed local Pegasus and Dashboard databases for "standalone" instances (e.g. adhoc instances) that have their own mysql instances.  
- recipe automation the manual `rake assets:precompile` step that was previously required.
- enhancements to add_frontend_server to invoke the "knife bootstrap" command with the appropriate role.

Below is a sample run.  

<code> <pre>
~/code-dot-org$  bin/add_frontend_server.rb -e adhoc -u phil -n phil-test16
Creating new adhoc instance using role unmonitored-standalone
Logging into gateway with username phil
Verifying connection to gateway
Using underscaled instance zone us-east-1b
Naming instance phil-test16, verifying that the name is okay
Name phil-test16 is okay, creating frontend server
Looking for instance id  i-a3842c01
Instance i-a3842c01 is now running, waiting for status checks to complete
Instance i-a3842c01 is healthy - adding to list of frontends for environment adhoc

Created instance i-a3842c01 with name phil-test16
Private dns name: ip-10-155-170-178.ec2.internal,
Public dns name: ec2-54-146-180-174.compute-1.amazonaws.com

Writing new configuration file
New configuration file uploaded, now loading it.
Updated Environment adhoc

Bootstrapping adhoc frontend, please be patient. This takes ~15 minutes.
ssh gateway -t knife bootstrap ip-10-155-170-178.ec2.internal -x ubuntu --sudo -E adhoc -N phil-test16 -r role[unmonitored-standalone]
## Connection to gateway.code.org closed.

Dashboard listening at: http://ec2-54-146-180-174.compute-1.amazonaws.com:8080
Pegasus listening at:   http://ec2-54-146-180-174.compute-1.amazonaws.com:8081
</pre></code>

### CODE REVIEW

**Comment by philbogle** [comment]
R: @mehalshah    Please review
CC: @wjordan @islemaster @trevorb  FYI

File: bin/add_frontend_server.rb [hunk 1]
```diff
@@ -220,5 +225,21 @@ def execute_ssh_on_channel(ssh, command, exit_error_string)
...
 end
 
-puts "To deploy new chef instance, run the following:"
-puts "ssh gateway -t knife bootstrap #{private_dns_name} -x ubuntu --sudo -E production -N #{instance_name} -r role[front-end]"
+case options['environment']
+when 'production'
+  cmd =  "ssh gateway -t knife bootstrap #{private_dns_name} -x ubuntu --sudo -E production -N #{instance_name} -r role[front-end]"
+when 'adhoc'
+  cmd =  "ssh gateway -t knife bootstrap #{private_dns_name} -x ubuntu --sudo -E adhoc -N #{instance_name} -r role[unmonitored-standalone]"
+else
```
Comment (mehalshah) [inline bin/add_frontend_server.rb]: Maybe make a map of environment->role consisting of "production=>front-end, adhoc=>unmonitored-standalone" so we can easily add more environments to this.

**Comment by islemaster** [comment]
This makes me happy! If launching an instance stays this easy, we'll be able to test and preview stuff much bettter!
:boom: :rocket: :star2:

File: bin/add_frontend_server.rb [hunk 2]
```diff
@@ -220,5 +225,21 @@ def execute_ssh_on_channel(ssh, command, exit_error_string)
...
+end
+
+if cmd
+  puts "Bootstrapping frontend: '#{cmd}'"
+  `#{cmd}`
+  puts '--------------------------------------------------------'
+  puts "Dashboard listening at: http://#{public_dns_name}:8080"
+  puts "Pegasus listening at:   http://#{public_dns_name}:8081"
+else
+  puts 'Unknown environment, not deploying'
```
Comment (mehalshah) [inline bin/add_frontend_server.rb]: Unknown environment, not bootstrapping?

**Comment by mehalshah** [comment]
LGTM - update the wiki when this deploys. We still have to do the load balancer stuff though, right?

**Comment by philbogle** [comment]
Yep, we still have to do the load balancer stuff manually for now.

File: bin/add_frontend_server.rb (hunk 1 above)
Comment (philbogle) [inline bin/add_frontend_server.rb]: Good suggestion, this simplifies this code by getting rid of the case statement and makes it easy to add new environments (once the recipes are good)

**Comment by laurelfan** [comment]
```
Developers can now fully bootstrap an adhoc or dashboard dashboard+pegasus instance
```

Do you mean all adhoc instances will be dashboard+pegasus or that you can optionally make an adhoc dashboard-only instance?

**Comment by philbogle** [comment]
I intended to say an adhoc or _production_ dashboard+pegasus instance.

File: cookbooks/cdo-apps/recipes/default.rb [hunk 3]
```diff
@@ -31,4 +31,24 @@
   user node[:current_user]
   group node[:current_user]
   action :nothing
+  notifies :run, 'execute[precompile-assets]', :immediately
+end
+
+execute "precompile-assets" do
+  command "rake assets:precompile"
```
Comment (wjordan) [inline cookbooks/cdo-apps/recipes/default.rb:38]: @philbogle should this command be prepended with `bundle exec`?

Comment (philbogle) [inline cookbooks/cdo-apps/recipes/default.rb:38]: Yes, good catch.  I submit a PR that changes that.

**Pull request merged**